### DISCUSSION TITLE
How to filter out rows that have `NULL` in _any_ column?

### DISCUSSION THREAD

**Discussion by emmansh** [duckdb/duckdb#5358] [Q&A]
Following [this question](https://stackoverflow.com/q/67076546/6105259) on StackOverflow, I wish to learn whether I could filter and keep only rows that have no `NULL` values in _any_ column.

Example
---
Let's simulate the following toy data:
```sql
-- duckdb
CREATE TABLE test (
  name varchar(20) NULL,
  lastname varchar(20) NULL
);
INSERT INTO test (name, lastname)
  VALUES (NULL, NULL),
  ('john', 'smith'),
  ('mike', NULL),
  (NULL, 'jones');
```

Such that:
```sql
-- duckdb
SELECT *
FROM test;
```
returns:
```
┌──────┬──────────┐
│ name ┆ lastname │
╞══════╪══════════╡
│      ┆          │
│ john ┆ smith    │
│ mike ┆          │
│      ┆ jones    │
└──────┴──────────┘
```
### My question
Is there a concise duckdb query to keep only rows that have no `NULL` values in any column? I've tried adopting the following Postgres syntax but it wouldn't work:
```sql
-- duckdb
SELECT *
FROM test
WHERE test.* IS NOT NULL; -- ✗
```
which returns the error:
> ```
> Parser Error: syntax error at or near "*"
> LINE 5: WHERE test.* IS NOT NULL;
>                    ^
> ```

Is there a duckdb equivalent or do I have to resort to specifiyng all column names in the table?
```sql
SELECT *
FROM test
WHERE name IS NOT NULL AND lastname IS NOT NULL; -- etc. etc. and if my table has 50 columns it's such a bummer
```

**Comment by Mytherin** [comment]
That's an interesting use case! We should add support for that postgres syntax and add support for using [`COLUMNS(*)`](https://duckdb.org/docs/sql/expressions/star#columns) in the `WHERE` clause.

**Reply by emmansh** [reply]
Thanks :)
In data analysis such step comes up often. In _R_ there's `tidyr::drop_na()` with the default behavior of dropping rows with any `NA` values. Similarly, `{dplyr}` has:
``` r
library(dplyr, warn.conflicts = FALSE)
library(tibble)

df <- tribble(~ id, ~ x, ~ y,
              1, 1, 0,
              2, 1, 1,
              3, NA, 1,
              4, 0, 0,
              5, 1, NA)

df |> 
  filter(!if_any(everything(), is.na))
#> # A tibble: 3 x 3
#>      id     x     y
#>   <dbl> <dbl> <dbl>
#> 1     1     1     0
#> 2     2     1     1
#> 3     4     0     0
```

**Comment by eitsupi** [comment]
I was trying to reproduce the contents of [Modern Polars](https://kevinheavey.github.io/modern-polars/) book with DuckDB and came across this discussion.
It would be wonderful if this could be implemented.

https://kevinheavey.github.io/modern-polars/tidy.html#cleaning

**Reply by Mytherin** [reply]
This has actually been implemented as part of #6621:

```sql
D SELECT * FROM (VALUES (NULL, 84), (42, NULL), (42, 84));
┌───────┬───────┐
│ col0  │ col1  │
│ int32 │ int32 │
├───────┼───────┤
│  NULL │    84 │
│    42 │  NULL │
│    42 │    84 │
└───────┴───────┘
D SELECT * FROM (VALUES (NULL, 84), (42, NULL), (42, 84)) WHERE COLUMNS(*) IS NOT NULL;
┌───────┬───────┐
│ col0  │ col1  │
│ int32 │ int32 │
├───────┼───────┤
│    42 │    84 │
└───────┴───────┘

```

**Reply by eitsupi** [reply]
Wow, thank you for your responce!
It seems unreleased versions feature? I'll try that!

**Comment by ericemc3** [comment]
to delete rows where all cols values are NULL:

FROM t
EXCEPT
FROM t
WHERE COLUMNS(*) IS NULL ;
